### ISSUE TITLE
Rule Request: `guard self` (vs weakSelf, strongSelf)

### ISSUE THREAD

**Issue by freak4pc** [realm/SwiftLint#2495]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

1. Why should this rule be added? 

Since Swift 4.2 we can now do `guard let self = self` instead of `guard let strongSelf`, `guad let weakSelf`, `guard let 'self'` and all other sort of oddities. I'd like to enforce that. 

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

Trigger:
```swift
if let `self` = self
guard let `self` = self
guard let strongSelf = self
guard let [anything not "self"] = self
```

Doesn't trigger:
```swift
if let self = self
guard let self = self
guard let self = [anytime this is self]
```

3. Should the rule be configurable, if so what parameters should be configurable?
I don't think so.

4. Should the rule be opt-in or enabled by default? Why?
Opt-in

**Comment by jpsim** [comment]
Thanks for the suggestion! I think this should be configurable on the rebinding identifier. There are still issues with Swift 4.2+ when rebinding `self` to `self`, especially around lldb (see [SR-6156](https://bugs.swift.org/browse/SR-6156)). For this reason, some projects have adopted a pattern to _always_ rebind `self` to something specifically different, in Lyft's case that's `guard let this = self`.

**Comment by freak4pc** [comment]
Yup, that makes sense :)

**Comment by lennartstolz** [comment]
I do feel like a grave desecrator - but I'd still love to see this rule becoming part of `SwiftLint`.

Especially because in the last ~3 yrs it almost came the standard in most codebases I'm working to use:

```swift
guard let self = self else { ... }
```

and the issue mentioned by @jpsim is also resolved by now. ✅ 

@marcelofabri : Do you think it's a valuable addition to SwiftLint? Is it "straightforward" to be implemented?
If so: I volunteer to implement it.

**Comment by dwirandytlvk** [comment]
Any update related to this?

**Comment by jpsim** [comment]
Adding a rule to do this in https://github.com/realm/SwiftLint/pull/4146

**Comment by samrayner** [comment]
Thank you for this @jpsim! Would it be possible to put the configured bind_identifier into the warning message so that devs know what identifier to use when they see the warning?
